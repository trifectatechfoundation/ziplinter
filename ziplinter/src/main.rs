use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
