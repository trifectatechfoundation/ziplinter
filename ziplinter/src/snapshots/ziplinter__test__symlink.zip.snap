---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "",
  "contents": [
    {
      "central": {
        "comment": "",
        "compressed_size": 9,
        "crc32": 2392783569,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2717843456,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          32,
          88,
          44,
          79,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 0,
        "internal_attrs": 0,
        "method": "Store",
        "mode": 134218239,
        "modified": "2012-02-03T21:56:48Z",
        "name": "symlink",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 9
      },
      "local": {
        "accessed": null,
        "compressed_size": 9,
        "crc32": 2392783569,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          32,
          88,
          44,
          79,
          32,
          88,
          44,
          79,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "gid": 1000,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2012-02-03T21:56:48Z",
        "name": "symlink",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 1000,
        "uncompressed_size": 9
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "comment": [],
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 1,
        "directory_offset": 74,
        "directory_records": 1,
        "directory_size": 77,
        "disk_nbr": 0
      },
      "offset": 151
    },
    "dir64": null,
    "global_offset": 0
  },
  "parsed_ranges": [
    {
      "end": 173,
      "kind": "eocd",
      "start": 151
    },
    {
      "end": 151,
      "kind": "cd header",
      "start": 74
    }
  ],
  "size": 173
}
